ragna/embedding_models/_all_minilm_l6_v2.py

```diff
@@ -0,0 +1,103 @@
+from collections import deque
+from typing import Deque, Iterable, Iterator, TypeVar, Union
+from uuid import UUID
+
+from ragna.core import (
+    Chunk,
+    Document,
+    Embedding,
+    EmbeddingModel,
+    PackageRequirement,
+    Page,
+    Requirement,
+)
+
+T = TypeVar("T")
+
+
+# The function is adapted from more_itertools.windowed to allow a ragged last window
+# https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.windowed
+def windowed_ragged(
+    iterable: Iterable[T], *, n: int, step: int
+) -> Iterator[tuple[T, ...]]:
+    window: Deque[T] = deque(maxlen=n)
+    i = n
+    for _ in map(window.append, iterable):
+        i -= 1
+        if not i:
+            i = step
+            yield tuple(window)
+
+    if len(window) < n:
+        yield tuple(window)
+    elif 0 < i < min(step, n):
+        yield tuple(window)[i:]
+
+
+class AllMiniLML6v2(EmbeddingModel):
+    @classmethod
+    def display_name(cls) -> str:
+        return "all-MiniLM-L6-v2"
+
+    @classmethod
+    def requirements(cls) -> list[Requirement]:
+        return [
+            # Rather than using sentence transformers package by itself, chromadb
+            # packaged this embedding model and is a much lighter dependency.
+            PackageRequirement("chromadb>=0.4.13"),
+            PackageRequirement("tiktoken"),
+        ]
+
+    def __init__(self) -> None:
+        super().__init__()
+        import tiktoken
+        from chromadb.utils import embedding_functions
+
+        self._tokenizer = tiktoken.get_encoding("cl100k_base")
+        self._model = embedding_functions.ONNXMiniLM_L6_V2()
+
+    def embed_documents(self, documents: list[Document]) -> list[Embedding]:
+        return [
+            Embedding(embedding=self.embed_text(chunk.text), chunk=chunk)
+            for document in documents
+            for chunk in self._chunk_pages(
+                document.extract_pages(),
+                document_id=document.id,
+                chunk_size=500,
+                chunk_overlap=250,
```

Comment:
These two need to become parameters of this function so users can still specify their values like they could before. Ultimately, we need to integrate this method into the Ragna protocol so that this can happen through the regular interface.

I can take that over when the PR is otherwise finished, because this will go deep into the Ragna internals. But LMK if you want me to guide you instead.

Reply:
resolved as for the time being and also permanently and indefinitely this will happen outside of the `EmbeddingModel`